torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1605,7 +1605,7 @@ c10::intrusive_ptr<Work> ProcessGroupNCCL::collective(
   work->opTimeout_ = options_->timeout;
   work->store_ = store_;
 
-  if (!coalescing_state_) {
+  if (!coalescing_state_ && capture_status == c10::cuda::CaptureStatus::None) {
     workEnqueue(work);
```

Comment:
I might prefer putting the capture status check inside `workEnqueue`, so that we could reject the enqueuing of any kind of collectives, P2P or coalesced collectives. Does that make sense?

Reply:
I think I haven't bypassed enough in all circumstances, hence the new test is giving an error in CI. I can try to fix that and to correct endCoalescing as well.

Reply:
Tests are now OK and I corrected endCoalescing. The check isn't inside workEnqueue though - I think it's simpler this way because in one case we already found the capture_status and in the other case we haven't, and also because it makes clear that more of the code could really be skipped when capturing.